include/mesh.tcc

```diff
@@ -347,6 +349,46 @@ bool mpm::Mesh<Tdim>::assign_velocity_constraints(
   return status;
 }
 
+//! Assign velocity constraints to cells
+template <unsigned Tdim>
+bool mpm::Mesh<Tdim>::assign_cell_velocity_constraints(
+    const std::vector<std::tuple<mpm::Index, unsigned, unsigned, double>>&
+        velocity_constraints) {
+  bool status = false;
+  try {
+    if (nodes_.size()) {
+      for (const auto& velocity_constraint : velocity_constraints) {
+        // Cell id
+        mpm::Index cell_id = std::get<0>(velocity_constraint);
+        // Face id
+        unsigned face_id = std::get<1>(velocity_constraint);
+        // Direction of the local coordinate system of the face
+        // Tdim = 2, Normal is y local axis, dir = 1
+        // Tdim = 3, Normal is z local axis, dir = 2
+        unsigned dir = std::get<2>(velocity_constraint);
+        // Velocity
+        double velocity = std::get<3>(velocity_constraint);
+
+        // Apply constraint
+        status = map_cells_[cell_id]->assign_velocity_constraint(face_id, dir,
+                                                                 velocity);
+
+        if (!status)
+          throw std::runtime_error(
+              "Cell or face or velocity constraint is invalid");
+      }
+    } else {
```

Comment:
Check and throw at the start, no need to nest the for loop inside a big if.

Reply:
Well there is a potential to throw within the for loop as well, thus we have the loop inside big if

Reply:
My question is why can't you just check `if (!nodes_.size())` and throw, why should the for loop be inside the `if` block?

Reply:
Yes but line 377 has another throw. It's either nest it or to have 2 `throw`, one for the `nodes_size()` and the other for `status` from `assign_velocity_constraint`

Reply:
Could you please explain why nesting a `for` loop inside an `if` helps, when both are inside a `try` block? If you move the `for` loop out of `if` that should help in branch prediction. What is the potential benefit of keeping this structure?

Reply:
Understood. My bad, I didn't pay attention to the `try` block